cartographer/mapping_2d/pose_graph.cc

```diff
@@ -369,8 +385,22 @@ void PoseGraph::WaitForAllComputations() {
 }
 
 void PoseGraph::FinishTrajectory(const int trajectory_id) {
-  // TODO(jihoonl): Add a logic to notify trimmers to finish the given
-  // trajectory.
+  AddWorkItem([this, trajectory_id]() REQUIRES(mutex_) {
+    const auto& submap_data = optimization_problem_.submap_data();
+    for (auto submap_id_data : submap_data.trajectory(trajectory_id)) {
+      submap_data_.at(submap_id_data.id).state = SubmapState::kFinished;
```

Comment:
This is what I meant: Wouldn't it be more consistent with the behavior of trimming submaps if we would
1. Tell the trimmer that we are interested in finishing a trajectory, i.e. this would not be part of the trimmer interface but a feature of the localization trimmer.
2. In `Trim()` implement this by telling the `trimming_handle` we would like to remove the trajectory. Then we also would not need a marker for the finished trajectory in the `PoseGraph` object.

Reply:
Um sorry I'm still confused what you meant by "implement this by telling the `trimming_handle` in `Trim()`. You meant add an additional variable in `TrimmingHandle` to indicate trajectories to remove? Then, query it in `Trim()`?

Reply:
I don't know when @wohe will be available again to continue the review, but we discussed this PR before he left, so I can take over.

The idea would be to add a method to the `Trimmable` interface to allow the `Trimmer` to query whether a trajectory is already finished. This method would have  to be implemented by the `TrimmingHandle` which just asks its parent `PoseGraph`. In `PoseGraph` I would implement the `FinishTrajectory` method as adding the trajectory_id to a set<int> finished_trajectories_ and implement a `PoseGraph::IsTrajectoryFinished()` method as `return finished_trajectories_.count(trajectory_id)`.

In the implementation of `PureLocalizationTrimmer::Trim(Trimmable*)` you can then check whether the trajectory the `PureLocalizationTrimmer` is responsible for is finished and if so, remove it.

Reply:
Um I thought @wohe don't want to have an additional marker(`finished_trajectories_`) in `PoseGraph` because he mentioned *we also would not need a marker for the finished trajectory in the PoseGraph object.*

I will update the PR as you explained. Let me know if anything changes.

Reply:
It was a bit of a back and forth when I discussed with him, but IIRC we settled on caching this information in the `PoseGraph` object.

Reply:
Ok.